This PR adds the pose_jacobian_derivative method in the DQ_Kinematics class.

juanjqo opened the pull request:
This PR adds the pose_jacobian_derivative method in the DQ_Kinematics class.

DQ_Kinematics has now the following two methods:

```cpp
    virtual MatrixXd pose_jacobian_derivative(const VectorXd& joint_configurations,
                                              const VectorXd& joint_velocity_configurations,
                                              const int& to_ith_link) const = 0;
    //Virtual methods
    virtual MatrixXd pose_jacobian_derivative(const VectorXd& joint_configurations,
                                              const VectorXd& joint_velocity_configurations) const;
```

DQ_SerialManipulator has now the following two methods:

```cpp
    virtual MatrixXd pose_jacobian_derivative(const VectorXd& q_vec, const VectorXd& q_vec_dot, const int& to_ith_link) const override; //Override from DQ_Kinematics
    virtual MatrixXd pose_jacobian_derivative(const VectorXd& q_vec, const VectorXd& q_vec_dot) const override; //Override from DQ_Kinematics
```

DQ_SerialManipulatorDH and DQ_SerialManipulatorMDH have now the following methods:

```cpp
    using DQ_SerialManipulator::raw_pose_jacobian_derivative;
    MatrixXd raw_pose_jacobian_derivative(const VectorXd& q_vec, const VectorXd& q_vec_dot, const int& to_ith_link) const override; 
```

DQ_DifferentialDriveRobot() has now the following method:

```cpp
    MatrixXd pose_jacobian_derivative (const VectorXd& q,
                                       const VectorXd& q_dot) const override;
```

DQ_HolonomicBase() has now the following methods:

```cpp
 virtual MatrixXd pose_jacobian_derivative(const VectorXd& q,
                                              const VectorXd& q_dot, const int& to_link) const override;
 virtual MatrixXd pose_jacobian_derivative(const VectorXd& q,
                                              const VectorXd& q_dot) const override

 MatrixXd raw_pose_jacobian_derivative(const VectorXd& q,
                                          const VectorXd& q_dot, const int& to_link = 2) const;
```
# Current limitations:

The classes DQ_SerialWholeBody() and DQ_WholeBody have not the pose_jacobian_derivative method available for the user.
If a user tries to use it the method raises an exception.

Example:

```cpp
MatrixXd DQ_WholeBody::pose_jacobian_derivative(const VectorXd &q, const VectorXd &q_dot, const int &to_ith_link) const
{
    throw std::runtime_error(std::string("pose_jacobian_derivative is not implemented yet."));
    return MatrixXd::Zero(1,1);
}
```

bvadorno commented:
Hi @juanjqo,

Thank you for the updated class diagram. Since `DQ_Kinematics` is an abstract class used with robots other than manipulators, it doesn't make much sense to use the variables `**joint**_configurations` and `**joint**_velocity_configurations` because some robots don't have joints.

I suggest using the variables either `robot_configurations` or simply `configurations` (and analogously for the velocities) in `DQ_Kinematics` and all subclasses of robots that don't have joints. I don't remember if we expose those variables by means of get and set methods. If we do, then we'd need to create aliases in `DQ_SerialManipulator` to maintain backward compatibility. 

Best wishes,
Bruno

juanjqo commented:
@dqrobotics/developers 

Hi Bruno, 

Thanks for your feedback. I implemented your suggestion. I renamed the variables in this way:

(For all classes except DQ_SerialManipulator)
```cpp
pose_jacobian_derivative(configurations, velocity_configurations, to_ith_link);
pose_jacobian_derivative(configurations, velocity_configurations);
```
Best regards, 

Juancho

mmmarinho commented:
@juanjqo Could we change from
`velocity_configurations`
to
`configuration_velocities`

or just stick to the `q` and `q_dot` we were using before?

Now it seems that all the subclasses have different parameter names for some reason.

I know that some are remnants of older versions of dqrobotics but we can take this change to unify their naming. Whatever we do, if we keep the naming consistent, it will be much easier to refactor when needed.

juanjqo commented:
@mmmarinho Sure! I personally like `q` and `q_dot` more than any other.  I just used the same names currently used by fkm(), pose_jacobian(), etc.

mmmarinho commented:
@juanjqo I’ll accept this one so the process on the Python version can also start.

Regardless of it being incorporated on the master branch, we (@bvadorno @juanjqo @mmmarinho ) can keep discussing and if necessary further refine this initial proposal.

juanjqo commented:
@dqrobotics/developers 

@mmmarinho sure!

I'm working on the Python version.
